play-json/shared/src/main/scala/play/api/libs/json/JsLookupResult.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.api.libs.json
+
+sealed trait JsLookupResult extends Any with JsReadable {
+  /**
+   * Tries to convert the node into a JsValue
+   */
+  def toOption: Option[JsValue]
+
+  def toEither: Either[JsonValidationError, JsValue]
+
+  def get: JsValue = toOption.get
+
+  def getOrElse(v: => JsValue): JsValue = toOption.getOrElse(v)
+  def isEmpty: Boolean
+  def isDefined: Boolean = !isEmpty
```

Comment:
`isDefined` and `getOrElse` could be `final` to enforce implementing class respect this contract

Reply:
The same could be done for `isEmpty`, and anything else that could be implemented based on other methods here. We should pick one style (either implementing in the trait or in subclasses) and stick to it. I'm not that concerned about forcing implementing classes to do things since we control the implementations.

Reply:
`isEmpty` and `isDefined` must follow the same logic, but be linked by negation, so having one depending on the type and the other always forced (`final` impl) makes sense to me, as in https://github.com/scala/scala/blob/2.12.x/src/library/scala/Option.scala#L107